scripts/build_osm_network.py

```diff
@@ -595,131 +595,97 @@ def create_station_at_equal_bus_locations(
     return lines, buses
 
 
-def _split_linestring_by_point(linestring, points):
+def fix_overpassing_lines(lines, buses, distance_crs, tol=1):
     """
-    Function to split a linestring geometry by multiple inner points.
+    Function to snap buses to lines that are within a certain tolerance.
 
     Parameters
     ----------
-    lstring : LineString
-        Linestring of the line to be split
-    points : list
-        List of points to split the linestring
-
-    Return
-    ------
-    list_lines : list
-        List of linestring to split the line
+    lines : GeoDataFrame
+        GeoDataFrame containing the lines
+    buses : GeoDataFrame
+        GeoDataFrame containing the buses
+    distance_crs : str
+        Coordinate reference system to use for distance calculations
+    tol : float
+        Tolerance in meters to snap the buses to the lines
+
+    Returns
+    -------
+    lines : GeoDataFrame
+        GeoDataFrame containing the lines
     """
 
-    list_linestrings = [linestring]
+    df_l = lines.copy()  # can use lines directly without copying
+    # drop all columns excpet id and geometry for buses
+    df_p = buses[["id", "geometry"]].copy()
 
-    for p in points:
-        # execute split to all lines and store results
-        temp_list = [split(l, p) for l in list_linestrings]
-        # nest all geometries
-        list_linestrings = [lstring for tval in temp_list for lstring in tval.geoms]
+    # change crs to distance based
+    df_l = df_l.to_crs(distance_crs)
+    df_p = df_p.to_crs(distance_crs)
 
-    return list_linestrings
+    # Buffer points to create areas for spatial join
+    buffer_df = gpd.GeoDataFrame(geometry=df_p.buffer(tol))
 
+    # Spatial join to find lines intersecting point buffers
+    joined = gpd.sjoin(df_l, buffer_df, how="inner", op="intersects")
 
-def fix_overpassing_lines(lines, buses, distance_crs, tol=1):
-    """
-    Function to avoid buses overpassing lines with no connection when the bus
-    is within a given tolerance from the line.
+    # group lines by their ids
+    group_lines = joined.groupby("id")
 
-    Parameters
-    ----------
-    lines : GeoDataFrame
-        Geodataframe of lines
-    buses : GeoDataFrame
-        Geodataframe of substations
-    tol : float
-        Tolerance in meters of the distance between the substation and the line
-        below which the line will be split
-    """
+    # iterate over the groups, TODO: change to apply
+    for i, group in group_lines:
+        line_id = group["id"].iloc[0]  # pick the line id that represents the group
```

Comment:
may line_id be the i of the forloop? probably not because of the different indexing, may be worth understanding if that makes sense

Reply:
good point, i is actually equivalent to line_id (in the next line).